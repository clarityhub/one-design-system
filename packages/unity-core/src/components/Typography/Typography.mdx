---
name: Typography
menu: Theme
---

import { Playground, PropsTable } from 'docz';
import Typography from './Typography';

# Typography

## \<Typography /\> component

```js
import Typography from '@clarityhub/unity-core/components/Typography';
```

## Prop Types

<PropsTable of={Typography} />

## Examples

### Text versus Paragraph

By default, if no `type` is provided to `Typography`, then `text` is used.

If you want to have multiple paragraphs of readable text, use `type="p"`.

<Playground>
    <Typography>Lorem ipsum dolor sit amet, <i>consectetur adipiscing elit</i>. Donec odio ipsum, <strong>sodales et ligula sed</strong>, venenatis cursus libero. Suspendisse sit amet tempus tortor, tincidunt egestas tortor. In hac habitasse platea dictumst. Sed consectetur nulla eget diam tincidunt eleifend.</Typography>
    <Typography type="p">Lorem ipsum dolor sit amet, <i>consectetur adipiscing elit</i>. Donec odio ipsum, <strong>sodales et ligula sed</strong>, venenatis cursus libero. Suspendisse sit amet tempus tortor, tincidunt egestas tortor. In hac habitasse platea dictumst. Sed consectetur nulla eget diam tincidunt eleifend.</Typography>
</Playground>

### Headings

<Playground>
    <Typography type="h1">H1. The quick brown fox</Typography>
    <Typography type="h2">H2. The quick brown fox</Typography>
    <Typography type="h3">H3. The quick brown fox</Typography>
    <Typography type="h4">H4. The quick brown fox</Typography>
    <Typography type="h5">H5. The quick brown fox</Typography>
</Playground>

Without padding:

<Playground>
    <Typography type="h1" noPadding>H1. The quick brown fox</Typography>
    <Typography type="h2" noPadding>H2. The quick brown fox</Typography>
    <Typography type="h3" noPadding>H3. The quick brown fox</Typography>
    <Typography type="h4" noPadding>H4. The quick brown fox</Typography>
    <Typography type="h5" noPadding>H5. The quick brown fox</Typography>
</Playground>

### Text

<Playground>
    <Typography type="p">Lorem ipsum dolor sit amet, <i>consectetur adipiscing elit</i>. Donec odio ipsum, <em>sodales et ligula sed</em>, venenatis cursus libero. Suspendisse sit amet tempus tortor, tincidunt egestas tortor. In hac habitasse platea dictumst. Sed consectetur nulla eget diam tincidunt eleifend.</Typography>
    <Typography type="p">Vestibulum quis facilisis dolor, <strong>a faucibus urna</strong>. Vestibulum lobortis massa in <b>lacus semper</b>, sit amet ultrices eros lacinia. Aenean consectetur luctus est, ut tincidunt arcu hendrerit eu. Aliquam orci dui, iaculis vitae vestibulum sed, ullamcorper ut odio. Maecenas ut mattis nisi. Fusce in ante sit amet magna ullamcorper euismod id ut lacus. Duis eget auctor velit.</Typography>
    <Typography type="subtext">This is a set of subtext</Typography>
    <Typography type="sectionLabel">This is a section label</Typography>
</Playground>

### Kitchen Sink

<Playground>
    <Typography type="h1">H1. The quick brown fox</Typography>
    <Typography type="p">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec odio ipsum, sodales et ligula sed, venenatis cursus libero. Suspendisse sit amet tempus tortor, tincidunt egestas tortor. In hac habitasse platea dictumst. Sed consectetur nulla eget diam tincidunt eleifend.</Typography>
    <Typography type="p">Vestibulum quis facilisis dolor, a faucibus urna. Vestibulum lobortis massa in lacus semper, sit amet ultrices eros lacinia. Aenean consectetur luctus est, ut tincidunt arcu hendrerit eu. Aliquam orci dui, iaculis vitae vestibulum sed, ullamcorper ut odio. Maecenas ut mattis nisi. Fusce in ante sit amet magna ullamcorper euismod id ut lacus. Duis eget auctor velit.</Typography>

    <Typography type="h2">H2. The quick brown fox</Typography>

    <Typography type="p">Vestibulum quis facilisis dolor, a faucibus urna. Vestibulum lobortis massa in lacus semper, sit amet ultrices eros lacinia. Aenean consectetur luctus est, ut tincidunt arcu hendrerit eu. Aliquam orci dui, iaculis vitae vestibulum sed, ullamcorper ut odio. Maecenas ut mattis nisi. Fusce in ante sit amet magna ullamcorper euismod id ut lacus. Duis eget auctor velit.</Typography>

    <Typography type="h2">H2. The quick brown fox</Typography>

    <Typography type="p">Vestibulum quis facilisis dolor, a faucibus urna. Vestibulum lobortis massa in lacus semper, sit amet ultrices eros lacinia. Aenean consectetur luctus est, ut tincidunt arcu hendrerit eu. Aliquam orci dui, iaculis vitae vestibulum sed, ullamcorper ut odio. Maecenas ut mattis nisi. Fusce in ante sit amet magna ullamcorper euismod id ut lacus. Duis eget auctor velit.</Typography>
</Playground>