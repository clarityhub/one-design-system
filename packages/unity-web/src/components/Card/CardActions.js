import styled from '@emotion/styled';

const CardActions = styled.div`
    margin-top: 1rem;
`;

export default CardActions;
