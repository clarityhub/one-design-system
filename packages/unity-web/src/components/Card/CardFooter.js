import styled from '@emotion/styled';
import { variants } from '@clarityhub/unity-core/lib/typography';
import borders from '@clarityhub/unity-core/lib/borders';
import colors from '@clarityhub/unity-core/lib/colors';

const CardFooter = styled.div`
    ${variants.text.string}

    color: ${colors.dark.default};

    padding: .75rem 1.25rem;
    border-top: 1px solid ${colors.muted.default};

    &:last-child {
        border-radius: 0 0 calc(${borders.borderRadius.rem}rem - 1px) calc(${borders.borderRadius.rem}rem - 1px);
    }
`;

export default CardFooter;
