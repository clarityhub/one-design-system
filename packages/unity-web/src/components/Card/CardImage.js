import styled from '@emotion/styled';
import borders from '@clarityhub/unity-core/lib/borders';

const CardImage = styled.img`
    display: block;
    margin: 0 auto;
    max-width: 100%;

    border-top-left-radius: calc(${borders.borderRadius.rem}rem - 1px);
    border-top-right-radius: calc(${borders.borderRadius.rem}rem - 1px);
`;

export default CardImage;
