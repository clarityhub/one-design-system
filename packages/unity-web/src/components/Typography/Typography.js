/** @jsx jsx */

import { bool, node, oneOf, string } from 'prop-types';
import styled from '@emotion/styled';
import { jsx } from '@emotion/core';
import colors from '@clarityhub/unity-core/lib/colors';
import { variants } from '@clarityhub/unity-core/lib/typography';

import { types } from '../../theme/fonts';

const noPaddingCss = `
    padding: 0;
`;

const noMarginCss = `
    margin-bottom: 0;
    margin-top: 0;

    & + & {
        margin-bottom: 0;
        margin-top: 0;
    }
`;

const Typography = ({
	type = 'text',
	variant,
	color = 'dark',
	children,
	component: suggestedComponent = 'p',
	noMargin = false,
	noPadding = false,
	inline = false,
	css: userCss,
	center = false,
	...rest
}) => {
	let styles = [
		Boolean(type) ? types[type]: {},
		Boolean(variant) ? variants[variant]: {},
		Boolean(color) ? { color: colors[color].default }: {},
		Boolean(noPadding) ? noPaddingCss: {},
		Boolean(noMargin) ? noMarginCss: {},
		Boolean(userCss) ? userCss: {},
		Boolean(inline) ? {
			display: 'inline-block',
		} : {},
		Boolean(center) ? {
			textAlign: 'center',
		} : {},
	];

	if (typeof suggestedComponent === 'string') {
		let component = suggestedComponent;
		if (suggestedComponent === 'p' && ['h1', 'h2', 'h3', 'h4', 'h5'].indexOf(type) !== -1) {
			component = type;
		}

		const Comp = styled[component](styles);
		return <Comp {...rest}>{children}</Comp>;
	} else {
		const Comp = suggestedComponent;
		return <Comp {...rest} css={styles}>{children}</Comp>;
	}
};

Typography.propTypes = {
	center: bool,
	children: node,
	color: oneOf(Object.keys(colors)),
	component: node,
	css: string,
	inline: bool,
	noMargin: bool,
	noPadding: bool,
	type: oneOf(Object.keys(types)),
	variant: oneOf(Object.keys(variants)),
};

export default Typography;
