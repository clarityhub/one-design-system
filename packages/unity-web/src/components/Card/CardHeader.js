import React from 'react';
import styled from '@emotion/styled';
import borders from '@clarityhub/unity-core/lib/borders';
import colors from '@clarityhub/unity-core/lib/colors';

import Typography from '../Typography';

const CardHeaderWrapper = styled.div`
    padding: .75rem 1.25rem;
    margin-bottom: 0;
    border-bottom: 1px solid ${colors.muted.default};

    &:first-of-type {
        border-radius: calc(${borders.borderRadius.rem}rem - 1px) calc(${borders.borderRadius.rem}rem - 1px) 0 0;
    }
`;

const CardHeader = ({ children, ...rest }) => (
	<CardHeaderWrapper {...rest}>
		<Typography noMargin noPadding type="h3">
			{children}
		</Typography>
	</CardHeaderWrapper>
);

export { CardHeaderWrapper };
export default CardHeader;
