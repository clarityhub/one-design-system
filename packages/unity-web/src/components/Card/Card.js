import styled from '@emotion/styled';
import isPropValid from '@emotion/is-prop-valid';
import { css } from '@emotion/core';
import borders from '@clarityhub/unity-core/lib/borders';
import colors from '@clarityhub/unity-core/lib/colors';
import { bool, oneOf } from 'prop-types';

import { CardHeaderWrapper } from './CardHeader';
import CardBody from './CardBody';

const Card = styled('div', {
	shouldForwardProp: prop => isPropValid(prop) && !['type'].includes(prop),
})(
	css`
        position: relative;
        display: flex;
        flex-direction: column;
        min-width: 0;
        word-wrap: break-word;
        background-clip: border-box;
        border-radius: ${borders.borderRadius.rem}rem;
    `,
	({ flat }) => !flat ? css`
        box-shadow: ${colors.shadow.default};
    ` : css`
        box-shadow: 0 0 0 1px rgba(0,0,0,0);
    `,
	({ type }) => {
		switch (type) {
		case 'notification':
			return css`
                background: ${colors.gradient.default};

                ${CardBody} {
                    padding-top: 2rem;
                    padding-bottom: 2rem;
                }
            `;
		case 'highlight':
			return css`
                background: ${colors.highlight.default};
            `;
		case 'callout':
			return css`
                background: #F9FAFC;
                border: 1px solid ${colors.primary.default};
            `;
		default:
			return css`
                background-color: ${colors.white.default};
            `;
		}
	},
	css`
        > :not(${CardHeaderWrapper}) {
            h3 {
                margin: 0;
                padding: 0;
            }

            h4 {
                margin-top: 0.6rem;
            }

            p:only-child {
                margin: 0;
            }
        }
    `,
	({ center }) => center && css`
        text-align: center;
    `,
	({ hoverable }) => hoverable && css`
        transition: box-shadow 0.2s ease-in-out;

        &:hover {
            box-shadow: 0 0 0 1px rgba(50,50,93,.1),
                0 7px 14px 0 rgba(50,50,93,.1),
                0 3px 6px 0 rgba(0,0,0,.07)
        }
    `,
);

/**
 * Hack for docz and react-docgen
 */
const CardPropTypes = () => <div />;
CardPropTypes.propTypes = {
	center: bool,
	flat: bool,
	hoverable: bool,
	type: oneOf(['notification', 'highlight', 'callout']),
};
CardPropTypes.defaultProps = {
	center: false,
	hoverable: false,
	flat: false,
};
Card.propTypes = CardPropTypes.propTypes;
export { CardPropTypes };

export default Card;
