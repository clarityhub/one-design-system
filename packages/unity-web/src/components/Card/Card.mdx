---
name: Card
menu: Components
---

import { Playground, PropsTable } from 'docz';
import Card, {
    CardHeader,
    CardFooter,
    CardImage,
    CardBody,
    CardActions,
} from './index';
import { CardPropTypes } from './Card';
import Typography from '../Typography';
import Button, { ButtonSet } from '../Buttons';
import Image from '../Image';

# Card

Cards provide a content container that you can use to display grouped content.

```js
import Card, {
    CardHeader,
    CardFooter,
    CardImage,
    CardBody,
    CardActions,
} from '@clarityhub/unity-web/lib/components/Card';
```

## Simple Card

In all examples, the headings and text use the `Typography` components.

<Playground>
    <Card>
        <CardBody>
            <Typography type="text">
                Default
            </Typography>
        </CardBody>
    </Card>
    <Card type="notification" style={{ marginTop: '1rem'}}>
        <CardBody>
            <Typography type="text" color="white">
                Notification
            </Typography>
        </CardBody>
    </Card>
</Playground>

## Card with Image

<Playground>
    <Card style={{ width: '16rem' }}>
        <CardImage src="https://placehold.it/300x150" />
        <CardBody>
            <Typography type="h3">Card title</Typography>
            <Typography type="text">
                Meow Mix
            </Typography>
        </CardBody>
    </Card>
</Playground>

## Card with Actions

<Playground>
    <Card>
        <CardBody>
            <Typography type="h3">Card title</Typography>
            <Typography type="h4">Card subtitle</Typography>
            <Typography type="text">
                Meow Mix
            </Typography>

            <CardActions>
                <ButtonSet>
                    <Button>Cancel</Button>
                    <Button type="danger">Delete</Button>
                </ButtonSet>
            </CardActions>
        </CardBody>
    </Card>
</Playground>

## Card with Header and Footer

<Playground>
    <Card>
        <CardHeader>
            Featured
        </CardHeader>
        <CardBody>
            <Typography type="h3">Card title</Typography>
            <Typography type="h4">Card subtitle</Typography>
            <Typography type="text">
                Meow Mix
            </Typography>

            <CardActions>
                <ButtonSet>
                    <Button>Cancel</Button>
                    <Button type="danger">Delete</Button>
                </ButtonSet>
            </CardActions>
        </CardBody>
        <CardFooter>
            2 days ago
        </CardFooter>
    </Card>
</Playground>

## Example Cards

<Playground>
    <Card center>
        <CardBody>
            <Typography type="h3">Welcome to Clarity Hub!</Typography>
            <Typography type="text">
                To get started, install the extension and use Intercom
                like you normally would. We’ll let you know when we
                have suggestions for you by providing a prompt in
                the place you normally work.
            </Typography>

            <Image
                center
                withMargin
                src="/public/get-started.png"
                alt="Get started with Clarity Hub"
                style={{ width: '80%' }}
            />

            <CardActions>
                <Button type="primary">Install the Extension</Button>
            </CardActions>
        </CardBody>
    </Card>
    <Card center style={{ marginTop: '1rem'}}>
        <CardBody>
            <Typography type="h3">Get Started</Typography>
            <Typography type="text">
                Send email invites to the rest of your team to get
                them onboarded! We’ll send them instructions
                on how to get started using Clarity Hub.
            </Typography>

            <CardActions>
                <Button type="primary">Invite Your Team</Button>
            </CardActions>
        </CardBody>
    </Card>
</Playground>

## Prop Types

The following prop types are available for the Card component.

<PropsTable of={CardPropTypes} />
