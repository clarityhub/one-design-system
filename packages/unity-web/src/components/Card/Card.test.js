import React, { Fragment } from 'react';
import renderer from 'react-test-renderer';

import Typography from '../Typography';
import Card, {
	CardBody,
} from './index';

describe('<Card />', () => {
	it('renders a simple card', () => {
		const node = (
			<Card>
				<CardBody>
					<Typography type="text">
                        Default
					</Typography>
				</CardBody>
			</Card>
		);
		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});
    
	it('renders different card types', () => {
		const node = (
			<Fragment>
				<Card type="notification">
					<CardBody>
						<Typography type="text" color="white">
                Notification
						</Typography>
					</CardBody>
				</Card>
				<Card flat type="highlight">
					<CardBody>
						<Typography type="text" color="dark">
                Highlight
						</Typography>
					</CardBody>
				</Card>
				<Card flat type="callout">
					<CardBody>
						<Typography type="text" color="dark">
                Callout
						</Typography>
					</CardBody>
				</Card>
			</Fragment>
		);

		const tree = renderer.create(node).toJSON();

		expect(tree).toMatchSnapshot();
	});
});