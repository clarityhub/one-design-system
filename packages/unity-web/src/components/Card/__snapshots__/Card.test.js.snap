// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> renders a simple card 1`] = `
<div
  className="css-hrz6r9-Card-Card-Card-Card-Card e6pferz0"
>
  <div
    className="css-opo7aj-CardBody em76frp0"
  >
    <p
      className="css-b04jnf"
    >
      Default
    </p>
  </div>
</div>
`;

exports[`<Card /> renders different card types 1`] = `
Array [
  <div
    className="css-8cmj7f-Card-Card-Card-Card-Card e6pferz0"
  >
    <div
      className="css-opo7aj-CardBody em76frp0"
    >
      <p
        className="css-1imar2r"
      >
        Notification
      </p>
    </div>
  </div>,
  <div
    className="css-1wi5bj8-Card-Card-Card-Card-Card e6pferz0"
  >
    <div
      className="css-opo7aj-CardBody em76frp0"
    >
      <p
        className="css-b04jnf"
      >
        Highlight
      </p>
    </div>
  </div>,
  <div
    className="css-1nzwfk4-Card-Card-Card-Card-Card e6pferz0"
  >
    <div
      className="css-opo7aj-CardBody em76frp0"
    >
      <p
        className="css-b04jnf"
      >
        Callout
      </p>
    </div>
  </div>,
]
`;
